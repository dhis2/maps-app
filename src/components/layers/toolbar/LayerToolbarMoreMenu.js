import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import i18n from '@dhis2/d2-i18n';
import { withStyles } from '@material-ui/core/styles';

import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Tooltip from '@material-ui/core/Tooltip';
import IconButton from '@material-ui/core/IconButton';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';

import MoreIcon from '@material-ui/icons/MoreHoriz';
import CreateIcon from '@material-ui/icons/Create';
import ViewListIcon from '@material-ui/icons/ViewList';
import DeleteIcon from '@material-ui/icons/Delete';
import SaveIcon from '@material-ui/icons/SaveAlt';

const styles = theme => ({
    button: {
        float: 'left',
        padding: 4,
        width: 32,
        height: 32,
    },
    divider: {
        margin: `${theme.spacing.unit}px 0`,
    },
});

export class LayerToolbarMoreMenu extends Component {
    state = {
        open: false,
        anchorEl: null,
    };

    static propTypes = {
        classes: PropTypes.object.isRequired,
        onEdit: PropTypes.func,
        onRemove: PropTypes.func,
        toggleDataTable: PropTypes.func,
        downloadData: PropTypes.func,
    };

    handleBtnClick = e => {
        this.setState({
            anchorEl: e.currentTarget,
            open: true,
        });
    };

    closeMenu = () => {
        this.setState({
            anchorEl: null,
            open: false,
        });
    };

    handleEditBtnClick = () => {
        this.closeMenu();
        this.props.onEdit();
    };

    handleDataTableBtnClick = () => {
        this.closeMenu();
        this.props.toggleDataTable();
    };

    handleRemoveBtnClick = () => {
        this.closeMenu();
        this.props.onRemove();
    };

    handleDownloadBtnClick = () => {
        this.closeMenu();
        this.props.downloadData();
    };

    render() {
        const {
            classes,
            onEdit,
            onRemove,
            toggleDataTable,
            downloadData,
        } = this.props;

        const somethingAboveDivider = toggleDataTable || downloadData,
            somethingBelowDivider = onRemove || onEdit,
            showDivider = somethingAboveDivider && somethingBelowDivider;

        if (!somethingAboveDivider && !somethingBelowDivider) {
            return null;
        }

        return (
            <Fragment>
                <Tooltip title={i18n.t('More actions')}>
                    <IconButton
                        className={classes.button}
                        onClick={this.handleBtnClick}
                    >
                        <MoreIcon />
                    </IconButton>
                </Tooltip>
                <Menu
                    anchorEl={this.state.anchorEl}
                    getContentAnchorEl={null}
                    anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
                    open={this.state.open}
                    onClose={this.closeMenu}
                    disableRestoreFocus={true} // Don't re-focus on the Tooltip after the dialog is closed
                >
                    {toggleDataTable && (
                        <MenuItem onClick={this.handleDataTableBtnClick}>
                            <ListItemIcon>
                                <ViewListIcon />
                            </ListItemIcon>
                            <ListItemText primary={i18n.t('Data table')} />
                        </MenuItem>
                    )}
                    {downloadData && (
                        <MenuItem onClick={this.handleDownloadBtnClick}>
                            <ListItemIcon>
                                <SaveIcon />
                            </ListItemIcon>
                            <ListItemText primary={i18n.t('Download data')} />
                        </MenuItem>
                    )}
                    {showDivider && (
                        <Divider className={classes.divider} light />
                    )}
                    {onEdit && (
                        <MenuItem onClick={this.handleEditBtnClick}>
                            <ListItemIcon>
                                <CreateIcon />
                            </ListItemIcon>
                            <ListItemText primary={i18n.t('Edit layer')} />
                        </MenuItem>
                    )}
                    {onRemove && (
                        <MenuItem onClick={this.handleRemoveBtnClick}>
                            <ListItemIcon>
                                <DeleteIcon />
                            </ListItemIcon>
                            <ListItemText primary={i18n.t('Remove layer')} />
                        </MenuItem>
                    )}
                </Menu>
            </Fragment>
        );
    }
}

export default withStyles(styles)(LayerToolbarMoreMenu);
