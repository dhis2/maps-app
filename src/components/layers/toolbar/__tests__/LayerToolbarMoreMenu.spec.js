import React from 'react';
import { shallow } from 'enzyme';
import { LayerToolbarMoreMenu } from '../LayerToolbarMoreMenu';

describe('LayerToolbarMoreMenu', () => {
    it('Should render nothing when no props passed', () => {
        const wrapper = shallow(<LayerToolbarMoreMenu classes={{}} />);
        expect(wrapper.equals(null)).toBe(true);
    });

    it('Should open menu on click', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu classes={{}} onRemove={() => null} />
        );

        expect(wrapper.state('open')).toBe(false);
        expect(wrapper.state('anchorEl')).toBe(null);
        wrapper.find('WithStyles(IconButton)').simulate('click', { currentTarget: 42 });
        expect(wrapper.state('open')).toBe(true);
        expect(wrapper.state('anchorEl')).toBe(42);
    });

    it('Should render a single MenuItem with no divider if the only prop is onRemove', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu classes={{}} onRemove={() => null} />
        );

        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1);

        expect(wrapper.find('WithStyles(MenuItem)').length).toBe(1);
        expect(wrapper.find('WithStyles(Divider)').length).toBe(0);
    });

    it('Should render two MenuItems with no divider if only passed onEdit and onRemove', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu classes={{}} onEdit={() => null} onRemove={() => null} />
        );

        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1);

        expect(wrapper.find('WithStyles(MenuItem)').length).toBe(2);
        expect(wrapper.find('WithStyles(Divider)').length).toBe(0);
    });

    it('Should render two MenuItems with no divider if only passed toggleDataTable', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu classes={{}} toggleDataTable={() => null} />
        );

        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1);

        expect(wrapper.find('WithStyles(MenuItem)').length).toBe(1);
        expect(wrapper.find('WithStyles(Divider)').length).toBe(0);
    });

    it('Should render two MenuItems with no divider if only passed toggleDataTable and downloadData', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu classes={{}} toggleDataTable={() => null} downloadData={() => null} />
        );

        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1);

        expect(wrapper.find('WithStyles(MenuItem)').length).toBe(2);
        expect(wrapper.find('WithStyles(Divider)').length).toBe(0);
    });

    it('Should render three MenuItems WITH divider if passed toggleDataTable, onEdit, and onRemove', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu classes={{}} onEdit={() => null} onRemove={() => null} toggleDataTable={() => null} />
        );

        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1);

        expect(wrapper.find('WithStyles(MenuItem)').length).toBe(3);
        expect(wrapper.find('WithStyles(Divider)').length).toBe(1);
    });

    it('Should render four MenuItems WITH divider if passed toggleDataTable, downloadData, onEdit, and onRemove', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu
                classes={{}}
                onEdit={() => null}
                onRemove={() => null}
                downloadData={() => null}
                toggleDataTable={() => null}
            />
        );

        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1);

        expect(wrapper.find('WithStyles(MenuItem)').length).toBe(4);
        expect(wrapper.find('WithStyles(Divider)').length).toBe(1);
    });

    it('Should match snapshot', () => {
        const wrapper = shallow(
            <LayerToolbarMoreMenu
                classes={{}}
                onEdit={() => null}
                onRemove={() => null}
                downloadData={() => null}
                toggleDataTable={() => null}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('Should respond to click events properly for MenuItems - toggleDataTable, downloadData, onEdit, and onRemove', () => {
        const onEdit = jest.fn(),
            onRemove = jest.fn(),
            toggleDataTable = jest.fn(),
            downloadData = jest.fn();

        const wrapper = shallow(
            <LayerToolbarMoreMenu
                classes={{}}
                onEdit={onEdit}
                onRemove={onRemove}
                downloadData={downloadData}
                toggleDataTable={toggleDataTable}
            />
        );

        wrapper
            .find('WithStyles(MenuItem)')
            .at(0)
            .simulate('click');
        expect(toggleDataTable).toHaveBeenCalledTimes(1);
        expect(downloadData).toHaveBeenCalledTimes(0);
        expect(onEdit).toHaveBeenCalledTimes(0);
        expect(onRemove).toHaveBeenCalledTimes(0);

        wrapper
            .find('WithStyles(MenuItem)')
            .at(1)
            .simulate('click');
        expect(toggleDataTable).toHaveBeenCalledTimes(1);
        expect(downloadData).toHaveBeenCalledTimes(1);
        expect(onEdit).toHaveBeenCalledTimes(0);
        expect(onRemove).toHaveBeenCalledTimes(0);

        wrapper
            .find('WithStyles(MenuItem)')
            .at(2)
            .simulate('click');
        expect(toggleDataTable).toHaveBeenCalledTimes(1);
        expect(downloadData).toHaveBeenCalledTimes(1);
        expect(onEdit).toHaveBeenCalledTimes(1);
        expect(onRemove).toHaveBeenCalledTimes(0);

        wrapper
            .find('WithStyles(MenuItem)')
            .at(3)
            .simulate('click');
        expect(toggleDataTable).toHaveBeenCalledTimes(1);
        expect(downloadData).toHaveBeenCalledTimes(1);
        expect(onEdit).toHaveBeenCalledTimes(1);
        expect(onRemove).toHaveBeenCalledTimes(1);
    });
});