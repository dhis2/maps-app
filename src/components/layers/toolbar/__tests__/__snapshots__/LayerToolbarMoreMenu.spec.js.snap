// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerToolbarMoreMenu Should match snapshot 1`] = `
<Fragment>
  <div
    className="moreMenuButton"
    data-test="moremenubutton"
    onClick={[Function]}
  >
    <IconButton
      dataTest="moremenubutton"
      onClick={[Function]}
      tooltip="More actions"
    >
      <SvgMore24 />
    </IconButton>
  </div>
  <Popover
    arrow={false}
    dataTest="dhis2-uicore-popover"
    elevation="0px 0px 1px rgba(33,41,52,0.1), 0px 4px 6px -1px rgba(33,41,52,0.1), 0px 2px 4px -1px rgba(33,41,52,0.06)"
    maxWidth={360}
    onClickOutside={[Function]}
    placement="right-start"
    reference={
      Object {
        "current": undefined,
      }
    }
  >
    <div
      className="menu"
    >
      <Menu
        dataTest="dhis2-uicore-menulist"
        dense={true}
      >
        <MenuItem
          dataTest="dhis2-uicore-menuitem"
          disabled={true}
          icon={<SvgTable16 />}
          label="Hide data table"
          onClick={[Function]}
        />
        <MenuItem
          dataTest="dhis2-uicore-menuitem"
          disabled={true}
          icon={<SvgDownload16 />}
          label="Download data"
          onClick={[Function]}
        />
        <Divider
          dataTest="dhis2-uicore-divider"
          margin="8px 0"
        />
        <MenuItem
          dataTest="dhis2-uicore-menuitem"
          icon={<SvgEdit16 />}
          label="Edit layer"
          onClick={[Function]}
        />
        <MenuItem
          dataTest="dhis2-uicore-menuitem"
          destructive={true}
          icon={<SvgDelete16 />}
          label="Remove layer"
          onClick={[Function]}
        />
      </Menu>
    </div>
  </Popover>
</Fragment>
`;
